### PR TITLE
feat(payments): CHECKOUT-6070 Add apple pay customer strategy

### PR DESCRIPTION
**Pull request by animesh1987**
## What?
Add apple pay customer strategy

## Why?
Add wallet button for apple pay, so shoppers can checkout using a wallet button in customer step.

## Testing / Proof
- Screencast

https://user-images.githubusercontent.com/7134802/146874158-8d39a584-d572-4f9f-b4e5-77117ee67673.mov

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        paymentMethod: PaymentMethod
+    ): ApplePayJS.ApplePayPaymentRequest {
+        const { storeProfile: { storeCountryCode, storeName } } = config;
+        const { currency: { code } } = cart;
+
+        const { initializationData : { merchantCapabilities, supportedNetworks } } = paymentMethod;
+
+        return {
+            requiredBillingContactFields: ['postalAddress'],
+            requiredShippingContactFields: ['email', 'phone', 'postalAddress'],
```
[the remark this answers is not in the document]
Comment (davidchin) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:147]: Also, phone number can be configured to be optional. So I believe we should respect the merchant's preference.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+            countryCode: storeCountryCode,
+            currencyCode: code,
+            merchantCapabilities,
+            supportedNetworks,
+            lineItems: [],
+            total: {
+                label: storeName,
+                amount: '0.01',
```
Comment (davidchin) [inline src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:155]: Why is the amount hardcoded?

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+        const { initializationData : { merchantCapabilities, supportedNetworks } } = paymentMethod;
+
+        return {
+            requiredBillingContactFields: ['postalAddress'],
+            requiredShippingContactFields: ['email', 'phone', 'postalAddress'],
+            countryCode: storeCountryCode,
+            currencyCode: code,
+            merchantCapabilities,
+            supportedNetworks,
+            lineItems: [],
```
Comment (davidchin) [inline src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:152]: Have you tested the flow where a shopper has filled in everything, and then goes back to the customer step and click on the ApplePay button. Would they be forced to start from the beginning?

File: src/customer/create-customer-strategy-registry.ts [hunk 4]
```diff
@@ -33,6 +39,7 @@ import { SquareCustomerStrategy } from './strategies/square';
 
 export default function createCustomerStrategyRegistry(
     store: CheckoutStore,
+    paymentClient: any,
```
[the remark this answers is not in the document]
Comment (animesh1987) [reply src/customer/create-customer-strategy-registry.ts:42]: I did try to find the type for this but can't find the type for this, it's been set to any in quite a few places. 

I can type it to use `ReturnType<typeof createPaymentClient>` which by inference resolves to `any` as well.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+        const container = document.getElementById(containerId);
+
+        if (!container) {
+            throw new InvalidArgumentError('Unable to create sign-in button without valid container ID.');
+        }
+
+        const button = document.createElement('input');
+        button.type = 'image';
+        button.src = this._paymentMethod?.logoUrl || '';
+        button.style.height = '40px';
```
[the remark this answers is not in the document]
Comment (animesh1987) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:112]: Well probably need feedback from product, I did think the same but also personally prefer we can get feedback on this.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (animesh1987) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:147]: Sorry this is the information that we are asking apple to return to us. Require word here means what we want apple to return to us.
Wording from their demo 
```
If you require address or contact details from your customer, request them by passing values in the requiredShippingContactFields or requiredBillingContactFields.
```

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 2 above)
Comment (animesh1987) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:155]: This is for initial display of sheet, it's pretty much apple being weird to ask for an amount. This information is not rendered. But we can't pass an empty string as well.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 3 above)
Comment (animesh1987) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:152]: This is one of the unresolved questions as to what we do when we close the apple pay modal. I did want to check with you that should we reset the checkout at this instance?

File: src/customer/create-customer-strategy-registry.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidchin) [reply src/customer/create-customer-strategy-registry.ts:42]: Right, I guess it's a JS library so it doesn't have any type. I guess don't worry about it unless you're keen to write the TS definition for it.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidchin) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:112]: Sure I don't mind having the size hardcoded at this stage.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidchin) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:147]: Right I see. In that case `postalAddress` should still be omitted if the cart contains digital items only.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 3 above)
Comment (davidchin) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:152]: What you described is a different flow to what I described. But yes, it's another flow we need to consider. I suppose it does make sense to reset the checkout details.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 2 above)
Comment (davidchin) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:155]: And you can't just provide the grand total of the cart?

File: src/customer/create-customer-strategy-registry.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (animesh1987) [reply src/customer/create-customer-strategy-registry.ts:42]: Probably will decouple this task, and do it as a separate task. Will be good to get an understanding of the that as well.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (animesh1987) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:147]: Cool will do that, I did think of that but wanted to check if it was a good place to do it here or do it as part of the shippingmethodselected callback.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 3 above)
Comment (animesh1987) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:152]: Will probably sync on this once we are back from holidays.

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 2 above)
Comment (animesh1987) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:155]: Yes, can do that too 😄

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts (hunk 3 above)
Comment (bc-peng) [reply src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:152]: In my dev store, when I goes back to the customer step by clicking `sign out`. `The shipping price you were quoted is no longer valid. Click OK to see the most up-to-date shipping prices` pops up. Then, everything reset.

https://user-images.githubusercontent.com/88361607/147194541-cbc0cc1f-48fe-4047-9c62-3e7ca5f6ac10.mov

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+        };
+
+        applePaySession.onshippingcontactselected = async event =>
+            this._handleShippingContactSelected(applePaySession, config, event);
+
+        applePaySession.onshippingmethodselected = async event =>
+            this._handleShippingMethodSelected(applePaySession, config, event);
+
+        applePaySession.oncancel = async () => {
+            await this._store.dispatch(this._remoteCheckoutActionCreator.signOut(paymentMethod.id));
```
Comment (davidchin) [inline src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:209]: What happens if these async calls fail in this callback? Should you trigger the `onError` handler, like what you're doing with `onvalidatemerchant`?

File: src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts [hunk 7]
```diff
@@ -206,9 +206,13 @@ export default class ApplePayCustomerStrategy implements CustomerStrategy {
...
 
         applePaySession.oncancel = async () => {
-            await this._store.dispatch(this._remoteCheckoutActionCreator.signOut(paymentMethod.id));
+            try {
+                await this._store.dispatch(this._remoteCheckoutActionCreator.signOut(paymentMethod.id));
 
-            return this._store.dispatch(this._checkoutActionCreator.loadCurrentCheckout());
+                return this._store.dispatch(this._checkoutActionCreator.loadCurrentCheckout());
+            } catch (error) {
+                return this._onError(error);
```
Comment (davidchin) [inline src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts:214]: This `error` instance would be an internal error. You'll need to map it to an external error class, like `PaymentMethodCancelledError`

**Review by davidchin — APPROVED** [review]
Nice work @animesh1987! 👍

**Pull request merged**